src/access/AccessRestriction.h

```diff
@@ -0,0 +1,288 @@
+     * @param [out] token        The unique token for the review, which can be matched to a review update event.
+     */
+    CHIP_ERROR RequestFabricRestrictionReview(FabricIndex fabricIndex, const std::vector<Entry> & arl, uint64_t & token)
+    {
+        token = ++mNextToken;
+        return DoRequestFabricRestrictionReview(fabricIndex, token, arl);
+    }
+
+    using EntryIterator = std::vector<SharedPtr<Entry>>::iterator;
+
+    /**
+     * Get iterator over the commissioning entries, which are used during commissioning
+     * and also installed as defaults for a fabric upon completion of commissioning.
+     *
+     * @param [out] begin iterator pointing to the beginning of the entries
+     * @param [out] end iterator pointing to the end of the entries
+     */
+    static CHIP_ERROR CommissioningEntries(EntryIterator & begin, EntryIterator & end)
+    {
+        begin = sCommissioningEntries.begin();
+        end   = sCommissioningEntries.end();
+
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * Get iterator over entries in the access restriction list by fabric.
+     *
+     * @param [in]  fabricIndex Iteration is confined to fabric
+     * @param [out] begin iterator pointing to the beginning of the entries
+     * @param [out] end iterator pointing to the end of the entries
+     */
+    CHIP_ERROR Entries(const FabricIndex fabricIndex, EntryIterator & begin, EntryIterator & end)
+    {
+        if (mFabricEntries.find(fabricIndex) == mFabricEntries.end())
+        {
+            return CHIP_ERROR_NOT_FOUND;
+        }
+
+        begin = mFabricEntries[fabricIndex].begin();
+        end   = mFabricEntries[fabricIndex].end();
+
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * Add a restriction entry to the commissioning access restriction list. This list is automatically
+     * applied to fabrics upon completion of commissioning. This request will fail if there is already
+     * an entry with the same cluster and endpoint.
+     *
+     * @param [in]  entry       The entry to add to the commissioning access restriction list.
+     * @return                  CHIP_NO_ERROR if the entry was successfully created, or an error code.
+     */
+    static CHIP_ERROR CreateCommissioningEntry(SharedPtr<Entry> entry);
+
+    /**
+     * Add a restriction entry to the access restriction list and notify any listeners. This request
+     * will fail if there is already an entry with the same cluster and endpoint.
+     *
+     * @param [out] index       (If not nullptr) index of created entry (relative to fabric).
+     * @param [in]  entry       The entry to add to the access restriction list.
+     * @param [in]  fabricIndex The index of the fabric in which the entry should be added.
+     * @return                  CHIP_NO_ERROR if the entry was successfully created, or an error code.
+     */
+    CHIP_ERROR CreateEntry(size_t * index, const Entry & entry, const FabricIndex fabricIndex);
```

Comment:
Suggest an overall `SetEntries`/`SetCommissioningEntries`.

This has the benefits of:

- Removals can be computed by set difference (if needed, would argue for removal of deletion events)
- Additions/updates the same
- You now know when the  list is done updating so you can call `MatterReportingAttributeChangeCallback`
- You can easily schedule a single setter lambda with a unique_ptr arg of the whole dataset to make sure it's called in correct context.

If you do this, you do have to normalize the list:

- The shadow copy/internal copy must always be kept in sorted order so that it's possible to tell the difference between changed and unchanged of doing SetXxx.
  - A unit test should be done that has out of order same entries and not report it.
  - That avoids over-reporting same data  (same semantic meaning) but different order because externally generated via things like unordered maps in the runtime)

Reply:
I believe @chrisdecenzo is going to remove the related entry event (updated/deleted).  I did implement the two new Set methods to change both lists completely.  I believe if the simplification is made to the spec, this should be resolved.